src/main/java/org/codehaus/groovy/runtime/memoize/StampedCommonCache.java

```diff
@@ -175,6 +174,11 @@ private V compute(K key, ValueProvider<? super K, ? extends V> valueProvider, bo
         return doWithReadLock(c -> c.values());
     }
 
+    @Override
+    public Set<Entry<K, V>> entrySet() {
+        return commonCache.entrySet();
```

Comment:
The code should be protected by `doWithReadLock`, we should do same thing for `isEmpty` `containsValue`, etc.

Reply:
true, but I think we have to throw UnsupportedOperationException for entrySet and keySet because they return a view which might then be modified under the covers without locks?

Reply:
Implementing an interface and throwing `UnsupportedOperationException` does not look elegant...

Reply:
Inelegant but the Java collections way! ;-) Better to look inelegant rather than provide an apparently good looking but buggy implementation! In any case, I have implemented entrySet() in the same way as keys() and values() are implemented. The concern though is that if you look at ConcurrentHashMap, keySet(), entrySet() and values() are backed by views using iterators and spliterators that guarantee "weak consistency". It doesn't seem like we are making any such considerations for StampedCommonCache? I will leave that as a future exercise - we can always create such views or throw UnsupportedOperationException if we deem there is no other way to be safe.

Reply:
I will set aside some time to complete the exercise 😉